domains/pallets/executive/src/lib.rs

```diff
@@ -402,55 +410,128 @@ where
+        let res = with_storage_layer(|| {
+            frame_executive::Executive::<
+                ExecutiveConfig,
+                BlockOf<ExecutiveConfig>,
+                Context,
+                UnsignedValidator,
+                AllPalletsWithSystem,
+                COnRuntimeUpgrade,
+            >::apply_extrinsic(uxt.clone())
+            .map_err(|err| DispatchError::Other(err.into()))
+        });
+
+        // apply extrinsic failed with transaction validity error
+        // this could happen for following scenarios
+        // - Bad extrinsic Signature
+        //      This extrinsic will be ignored by the operators during the bundle check
+        //      and marks such bundle as Invalid.
+        // - Extrinsic execution failed
+        //      There are multiple scenarios why this can happen
+        //      - Inherent extrinsic failed. If this happens, we should fail to apply extrinsic
+        //      - Pre and Post dispatch fails. Check the test `test_domain_block_builder_include_ext_with_failed_predispatch`
+        //        why this could happen. If it fail due to this, then we revert the inner storage changes
+        //        but still include extrinsic so that we can clear inconsistency between block body and trace roots.
+        let res = match res {
+            Ok(dispatch_outcome) => Ok(dispatch_outcome),
+            Err(err) => {
+                let encoded = uxt.encode();
+                let (maybe_signer, dispatch_info) =
+                    ExecutiveConfig::ExtrinsicStorageFees::extract_signer(uxt);
+                // if this is not a normal extrinsic, then transaction should not execute
+                // we should fail here.
+                if dispatch_info.class != DispatchClass::Normal {
+                    return Err(TransactionValidityError::Invalid(
+                        InvalidTransaction::BadMandatory,
+                    ));
+                }
+
+                // charge signer for extrinsic storage
+                // TODO: charge user for the storage of extrinsic on consensus chain
+                if let Some(signer) = maybe_signer {
+                    let storage_fees = ExecutiveConfig::LengthToFee::weight_to_fee(
+                        &Weight::from_parts(encoded.len() as u64, 0),
+                    );
+
+                    // best effort to charge the fees to signer.
+                    // if signer does not have enough balance, we continue
+                    let maybe_charged_fees = ExecutiveConfig::Currency::burn_from(
+                        &signer,
+                        storage_fees,
+                        Precision::BestEffort,
+                        Fortitude::Force,
+                    );
+
+                    if let Ok(charged_fees) = maybe_charged_fees {
+                        ExecutiveConfig::ExtrinsicStorageFees::on_storage_fees_charged(charged_fees)
+                    }
```

Comment:
Can we use `pallet_transaction_payment::OnChargeTransaction` as usual?

Reply:
explained above